Bump version to 0.0.1a27 (#1586)
astral-sh/ty: CHANGELOG.md

- Support `isinstance()` and `issubclass()` narrowing when the second argument is a `typing.py` stdlib alias ([#21391](https://github.com/astral-sh/ruff/pull/21391))
- Support `type[…]` and `Type[…]` in implicit type aliases ([#21421](https://github.com/astral-sh/ruff/pull/21421))
- Support attribute-expression `TYPE_CHECKING` conditionals ([#21449](https://github.com/astral-sh/ruff/pull/21449))
- Support class-arguments for dataclass transformers ([#21457](https://github.com/astral-sh/ruff/pull/21457))
- Support legacy `typing` special forms in implicit type aliases ([#21433](https://github.com/astral-sh/ruff/pull/21433))
- Support stringified annotations in value-position `Annotated` instances ([#21447](https://github.com/astral-sh/ruff/pull/21447))
- Support all parameters of dataclass transforms ([#21474](https://github.com/astral-sh/ruff/pull/21474))
- Support `__hash__` semantics and `unsafe_hash` for dataclasses ([#21470](https://github.com/astral-sh/ruff/pull/21470))
- Improve handling of version-specific features of dataclasses ([#21453](https://github.com/astral-sh/ruff/pull/21453))

> - Support annotated `self` parameter in constructor of non-invariant generic classes ([#21325](https://github.com/astral-sh/ruff/pull/21325))

Reviewer: ```suggestion
- Correctly infer the specialization of a non-invariant PEP-695 generic class that has an annotated `self` parameter in its `__init__` method ([#21325](https://github.com/astral-sh/ruff/pull/21325))
```

- Improve generic constructor inference ([#20933](https://github.com/astral-sh/ruff/pull/20933))
- Improve generic call expression inference ([#21210](https://github.com/astral-sh/ruff/pull/21210))
- Improve generic class constructor inference ([#21442](https://github.com/astral-sh/ruff/pull/21442))

> - Improve literal promotion heuristics ([#21439](https://github.com/astral-sh/ruff/pull/21439))

Reviewer: ```suggestion
- Improve heuristics used to decide when it is appropriate to "promote" a `Literal` type such as `Literal[42]` to its instance supertype (in this case, `int`) when solving type variables ([#21439](https://github.com/astral-sh/ruff/pull/21439))
```

- Consider `from thispackage import y` a re-export of `y` in `__init__.pyi` ([#21387](https://github.com/astral-sh/ruff/pull/21387))
- Allow PEP-604 unions in stubs and `TYPE_CHECKING` blocks prior to 3.10 ([#21379](https://github.com/astral-sh/ruff/pull/21379))
- Ensure annotation/type expressions in stub files are always deferred ([#21401](https://github.com/astral-sh/ruff/pull/21401))
- Further improve details around which expressions should be deferred in stub files ([#21456](https://github.com/astral-sh/ruff/pull/21456))
- Sync vendored typeshed stubs ([#21466](https://github.com/astral-sh/ruff/pull/21466))

### LSP server

- Support for notebooks in VS Code ([#21175](https://github.com/astral-sh/ruff/pull/21175))
- Fix goto for `float` and `complex` in type annotation positions ([#21388](https://github.com/astral-sh/ruff/pull/21388))
- Add more keywords to scope-based completions ([#21383](https://github.com/astral-sh/ruff/pull/21383))
- Add synthetic members to completions on dataclasses ([#21446](https://github.com/astral-sh/ruff/pull/21446))
- Provide `import` completion when in `from <name> <name>` statement ([#21291](https://github.com/astral-sh/ruff/pull/21291))

> - Suppress completions when introducing names with `as` ([#21460](https://github.com/astral-sh/ruff/pull/21460))

Reviewer: ```suggestion
- Suppress completion suggestions following `as` tokens ([#21460](https://github.com/astral-sh/ruff/pull/21460))
```

- Suppress invalid suggestions in import statements ([#21484](https://github.com/astral-sh/ruff/pull/21484))

> - Improve semantic token classification for names ([#21399](https://github.com/astral-sh/ruff/pull/21399))

Reviewer: ```suggestion
- Improve [semantic token](https://microsoft.github.io/language-server-protocol/specifications/lsp/3.17/specification/#textDocument_semanticTokens) classification for names ([#21399](https://github.com/astral-sh/ruff/pull/21399))
```

- Set `definition` modifier for parameter declarations when computing semantic tokens ([#21420](https://github.com/astral-sh/ruff/pull/21420))

> - Inlay hint call argument location ([#20349](https://github.com/astral-sh/ruff/pull/20349))

Reviewer 2: Without looking at the PR, I'm not clear from this description what this actually does

> - Inlay hint call argument location ([#20349](https://github.com/astral-sh/ruff/pull/20349))

Reviewer: ```suggestion
- Enable goto-definition when double-clicking on an inlay hint ([#20349](https://github.com/astral-sh/ruff/pull/20349))
```
